server/api/src/rsDataObjRepl.cpp

```diff
@@ -479,60 +483,51 @@ namespace
 
-        // If the user has no permissions on the object, it may have returned an empty result set.
-        if (permission_query.empty()) {
-            return false;
+        constexpr auto minimum_permission_required_for_replication = access_type::modify_object;
+
+        // Check access permission for the user first...
+        const auto user_permission_query_string =
+            fmt::format("select DATA_ACCESS_TYPE where USER_NAME = '{}' and USER_ZONE = '{}' and DATA_ID = '{}' and "
+                        "DATA_ACCESS_TYPE >= '{}'",
+                        user.name,
+                        user.zone,
+                        _obj->data_id(),
+                        static_cast<int>(minimum_permission_required_for_replication));
+        auto user_permission_query = irods::query{&_comm, user_permission_query_string};
+        if (!user_permission_query.empty()) {
+            return true;
         }
 
-        const auto& query_result = permission_query.front();
-        const auto& access_type_string = query_result[0];
-
-        try {
-            using access_type = irods::experimental::catalog::access_type;
-
-            switch (std::stoi(access_type_string)) {
-                case static_cast<int>(access_type::modify_object): // NOLINT(bugprone-branch-clone)
-                    [[fallthrough]];
-                case static_cast<int>(access_type::delete_object): // NOLINT(bugprone-branch-clone)
-                    [[fallthrough]];
-                case static_cast<int>(access_type::own):
-                    return true;
-
-                default:
-                    return false;
-            }
-        }
-        catch (const std::invalid_argument& e) {
-            log_api::error("DATA_ACCESS_TYPE for user [{}#{}] on object [{}] has invalid value [{}]",
-                           _comm.clientUser.userName,
-                           _comm.clientUser.rodsZone,
-                           _obj->logical_path(),
-                           access_type_string);
-            return false;
-        }
-        catch (const std::out_of_range& e) {
-            log_api::error("DATA_ACCESS_TYPE for user [{}#{}] on object [{}] has out-of-range value [{}]",
-                           _comm.clientUser.userName,
-                           _comm.clientUser.rodsZone,
-                           _obj->logical_path(),
-                           access_type_string);
-            return false;
-        }
+        // If the user does not have sufficient permissions on the data object for replication, perhaps one of their
+        // groups does...
+        const auto groups = adm::server::groups(_comm, user);
+        std::vector<std::string> groups_with_quotes;
+        groups_with_quotes.reserve(groups.size());
+        std::for_each(groups.cbegin(), groups.cend(), [&groups_with_quotes](const auto& group) -> void {
+            groups_with_quotes.emplace_back(fmt::format("'{}'", group.name));
+        });
```

Comment:
I wasn't quite sure how to get that work here. Can you provide a rough sketch of what that would look like?

Reply:
I think it would be:
```cpp
std::transform(std::cbegin(groups), std::cend(groups), std::back_inserter(groups_with_quotes), [](const auto& val) -> std::string {
  return fmt::format("'{}'", val);
});
```

Didn't try it though.

Reply:
> I think it would be:
> 
> ```c++
> std::transform(std::cbegin(groups), std::cend(groups), std::back_inserter(groups_with_quotes), [](const auto& val) -> std::string {
>   return fmt::format("'{}'", val);
> });
> ```
> 
> Didn't try it though.

Yeah, what Martin wrote looks right to me. Alternatively, if you want to avoid the auto resize with the vector, I think you can resize and then pass in the destination iterator.

```cpp
groups_with_quotes.reserve(groups.size());
std::transform(std::cbegin(groups), std::cend(groups), std::begin(groups_with_quotes), [](const auto& val) {
  return fmt::format("'{}'", val);
});
```
The benefit of this one is that I think a parallel implementation of `std::transform` can work better (though I doubt that'll ever be an issue...)

Reply:
Ahh, this is actually just what I was looking for. I'll try this out and report back.

Reply:
Given that this line is just putting apostrophes around some strings for a query (inexpensive relative to running the query), I don't think we need to worry about parallel implementations. After some discussion, I think the `std::back_inserter` solution will work fine here. I've created a separate issue for discussion in our copious free time for a deeper analysis of the approaches: https://github.com/irods/irods/issues/7832

And for future readers... The destination `std::vector` used with `std::transform` needs to have its elements constructed ahead of time using `resize()` rather than `reserve()`!

Reply:
Agreed. Good investigation.

Reply:
I've confirmed that the `std::back_inserter` / `std::vector::reserve` solution works, so I pushed that up. Thanks for the suggestion and analysis so far. Resolving.